### PR TITLE
Support security posture config

### PR DESCRIPTION
**Pull request by DDDDarrenWB**
Add support of GKE Security Posture Config.

1. Add `security_posture_config.mode` field in schema.
2. Add `security_posture_config.vulnerability_mode` field in schema.

They both talks to GKE API - Security Posture Config. 
The behavior of terraform matches with how user will use `gcloud` to manage our feature.

Fixes https://github.com/hashicorp/terraform-provider-google/issues/14973

If this PR is for Terraform, I acknowledge that I have:

- [X ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: added field `security_posture_config` to resource `google_container_cluster`
```

### CODE REVIEW

**Comment by DDDDarrenWB** [comment]
The failed test seems not related to my change? My changes are tested only in `TestAccContainerCluster_withSecurityPostureConfig`.

I have also updated the release note.

**Comment by zli82016** [comment]
> The failed test seems not related to my changes? My changes are tested only in `TestAccContainerCluster_withSecurityPostureConfig`.
> 
> I have also updated the release note.

Right, the failed test is not related to this PR.

**Comment by DDDDarrenWB** [comment]
Thanks @zli82016 Anything else I should do at this point?

**Comment by zli82016** [comment]
> Thanks @zli82016 Anything else I should do at this point?

Thanks for working on it. Not now. I will review it and give you update soon.

**Comment by zli82016** [comment]
Hello, @DDDDarrenWB , I am a little confused about this PR. I may miss something.

In the API for the resource cluster, I do not see the two new fields. https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters

Also, https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/1586 is already closed. In this issue, it requested to add `protect_config` instead of the two new fields `security_posture` and `workload_vulnerability_scanning`

**Comment by DDDDarrenWB** [comment]
Hi @zli82016 

1. The new fields are documented here, https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#securitypostureconfig

2. You are right. That is issue is for `protect_config`. I opened a new issue for this PR, https://github.com/hashicorp/terraform-provider-google/issues/14973. Thank you!

**Comment by zli82016** [comment]
> Hi @zli82016
> 
> 1. The new fields are documented here, https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#securitypostureconfig
> 2. You are right. That is issue is for `protect_config`. I opened a new issue for this PR, [Add support for SecurityPostureConfig hashicorp/terraform-provider-google#14973](https://github.com/hashicorp/terraform-provider-google/issues/14973). Thank you!

Thanks for the information. 
Instead of adding these two fields to the top level of the schema, it is better to wrap them into the field `security_posture_config`, which is similar to [`protect_config`]( https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb#L1050)

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb [hunk 1]
```diff
@@ -1090,6 +1090,20 @@ func ResourceContainerCluster() *schema.Resource {
 			},
 <% end -%>
 
+			"security_posture": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1096]: I wonder what is the reason to set `Computed: true` here? `Computed:     true` is used in the case that the API will returns default value if no value is specified in the request.

Comment (DDDDarrenWB) [reply mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1096]: Good question! I couldn't pass the Acceptance test locally if I don't set the `Computed:true`. Although, it didn't return error when I did manual test without `Computed:true`. I really could use your advice on this. 

Could it be the configurations I have in my local development environment?

**Comment by DDDDarrenWB** [comment]
> > Hi @zli82016
> > 
> > 1. The new fields are documented here, https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#securitypostureconfig
> > 2. You are right. That is issue is for `protect_config`. I opened a new issue for this PR, [Add support for SecurityPostureConfig hashicorp/terraform-provider-google#14973](https://github.com/hashicorp/terraform-provider-google/issues/14973). Thank you!
> 
> Thanks for the information. Instead of adding these two fields to the top level of the schema, it is better to wrap them into the field `security_posture_config`, similar to [`protect_config`](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb#L1050)

I want to make the behavior of terraform matches to how user will use `gcloud` to manage our feature. 
We offer two flags for user to enable our features via gcloud, `--workload-vulnerability-scanning` and `--security-posture`. 

If the rule of Terraform is to make it similar to API, I'll also be happy to change my codes.

**Comment by zli82016** [comment]
> > > Hi @zli82016
> > > 
> > > 1. The new fields are documented here, https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#securitypostureconfig
> > > 2. You are right. That is issue is for `protect_config`. I opened a new issue for this PR, [Add support for SecurityPostureConfig hashicorp/terraform-provider-google#14973](https://github.com/hashicorp/terraform-provider-google/issues/14973). Thank you!
> > 
> > 
> > Thanks for the information. Instead of adding these two fields to the top level of the schema, it is better to wrap them into the field `security_posture_config`, similar to [`protect_config`](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb#L1050)
> 
> I want to make the behavior of terraform matches to how user will use `gcloud` to manage our feature. We offer two flags for user to enable our features via gcloud, `--workload-vulnerability-scanning` and `--security-posture`.
> 
> If the rule of Terraform is to make it similar to API, I'll also be happy to change my codes.

I see. In Terraform, we will make it similar to API.

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb (hunk 1 above)
Comment (zli82016) [reply mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1096]: Did you get any error in the logs when running the acceptance test locally without `Computed:true`?

**Comment by zli82016** [comment]
Also, in Terraform, the field name will match the name in the API. For example, the field will look like:
```
security_posture_config {
   mode               = "DISABLED"
   vulnerability_mode = "DISABLED"
}
```

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb (hunk 1 above)
Comment (DDDDarrenWB) [reply mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1096]: Yes.

Error log like below,
```
error=
  | After applying this test step, the plan was not empty.
  | stdout:
  | 
  | 
  | Terraform used the selected providers to generate the following execution
  | plan. Resource actions are indicated with the following symbols:
  |   ~ update in-place
  | 
  | Terraform will perform the following actions:
  | 
  |   # google_container_cluster.with_security_posture_config will be updated in-place
  |   ~ resource "google_container_cluster" "with_security_posture_config" {
  |         id                              = "projects/wangbei-gke-dev/locations/us-central1-c/clusters/tf-test-cluster-6mwho946jg"
  |         name                            = "tf-test-cluster-6mwho946jg"
  |       - workload_vulnerability_scanning = "DISABLED" -> null
  |         # (25 unchanged attributes hidden)
  | 
  |         # (16 unchanged blocks hidden)
  |     }
  | 
  | Plan: 0 to add, 1 to change, 0 to destroy.
```

**Comment by DDDDarrenWB** [comment]
> Also, in Terraform, the field name will match the name in the API. For example, the field will look like:
> 
> ```
> security_posture_config {
>    mode               = "DISABLED"
>    vulnerability_mode = "DISABLED"
> }
> ```

Understood. I'll make the change.

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb (hunk 1 above)
Comment (zli82016) [reply mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1096]: I think the `Computed: true` is needed.

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb [hunk 2]
```diff
@@ -1090,6 +1090,20 @@ func ResourceContainerCluster() *schema.Resource {
...
+				ValidateFunc: validation.StringInSlice([]string{"DISABLED", "STANDARD"}, false),
+				Description:  `Sets the mode of the Kubernetes security posture API's off-cluster features. Available options include DISABLED and STANDARD.`,
+			},
+			"workload_vulnerability_scanning": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
```
[the remark this answers is not in the document]
Comment (zli82016) [reply mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:1103]: Computed: true is needed. Sorry for the confusion.

**Comment by DDDDarrenWB** [comment]
reopen due to false close.

File: mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb [hunk 3]
```diff
@@ -2299,6 +2325,10 @@ func resourceContainerClusterCreate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
+	if v, ok := d.GetOk("security_posture_config"); ok {
+		cluster.SecurityPostureConfig = expandSecurityPostureConfig(v)
+	}
+
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb:2331]: The code block should be put before sending out CREATE request, like line 2230.

**Comment by zli82016** [comment]
I run the test `TestAccContainerCluster_withSecurityPostureConfig` in the teamcity and it passed.

https://ci-oss.hashicorp.engineering/viewLog.html?buildId=412914&buildTypeId=GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream&tab=testsInfo

**Review by zli82016 — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**